cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -0,0 +1,304 @@
+                }),
+            });
+            appRunnerClient = new AWS.AppRunner();
+
+            switch (event.RequestType) {
+                case "Create":
+                    await addCustomDomain(serviceARN, customDomain);
+                    await waitForCustomDomainToBeActive(serviceARN, customDomain);
+                    physicalResourceID = `/associate-domain-app-runner/${customDomain}`;
+                    break;
+                case "Update":
+                case "Delete":
+                    throw new Error("not yet implemented");
+                default:
+                    throw new Error(`Unsupported request type ${event.RequestType}`);
+            }
+            await report(event, context, "SUCCESS", physicalResourceID);
+        } catch (err) {
+            if (err.name === ERR_NAME_INVALID_REQUEST && err.message.includes(`${customDomain} is already associated with`)) {
+                await report(event, context, "SUCCESS", physicalResourceID);
+                return;
+            }
+            console.log(`Caught error for service ${serviceARN}: ${err.message}`);
+            await report(event, context, "FAILED", physicalResourceID, null, err.message);
+        }
+    };
+
+    await Promise.race([lambdaTimedOut, handler(),]);
+};
+
+/**
+ * Wait for the custom domain to be ACTIVE.
+ * @param {string} serviceARN the service to which the domain is added.
+ * @param {string} customDomainName the domain name.
+ */
+async function waitForCustomDomainToBeActive(serviceARN, customDomainName) {
+    let i;
+    for (i = 0; i < ATTEMPTS_WAIT_FOR_ACTIVE; i++) {
+        const data = await appRunnerClient.describeCustomDomains({
+            ServiceArn: serviceARN,
+        }).promise().catch(err => {
+            throw new Error(`wait for domain ${customDomainName} to be active: ` + err.message);
+        });
+
+        let domain;
+        for (const d of data.CustomDomains) {
+            if (d.DomainName === customDomainName) {
+                domain = d;
+                break;
+            }
+        }
+
+        if (!domain) {
+            throw new Error(`wait for domain ${customDomainName} to be active: : domain ${customDomainName} is not associated`);
+        }
+
+        if (domain.Status !== DOMAIN_STATUS_ACTIVE) {
+            // Exponential backoff with jitter based on 200ms base
+            // component of backoff fixed to ensure minimum total wait time on
+            // slow targets.
+            const base = Math.pow(2, i);
+            await sleep(Math.random() * base * 50 + base * 150);
+            continue;
+        }
+        return;
+    }
+
+    if (i === ATTEMPTS_WAIT_FOR_ACTIVE) {
+        console.log("Fail to wait for state to become ACTIVE. However, this doesn't necessarily mean the operation has failed. It usually takes a long time to validate domain and can be longer than the 15 minutes duration for which a Lambda function can run at most.");
```

Comment:
> However, this doesn't necessarily mean the operation has failed.

But we will rollback and trigger the lambda again to potentially undo everything. What do you think of:

> Fail to wait for the domain status to become ACTIVE. It usually takes a long time to validate domain and can be longer than the 15 minutes duration for which a Lambda function can run at most. Try associating the domain manually.